README.md
# Nuxt Responsive Loader

[![npm (scoped with tag)](https://img.shields.io/npm/v/nuxt-responsive-loader/latest.svg?style=flat-square)](https://npmjs.com/package/nuxt-responsive-loader)
[![npm](https://img.shields.io/npm/dt/nuxt-responsive-loader.svg?style=flat-square)](https://npmjs.com/package/nuxt-responsive-loader)

>

[📖 **Release Notes**](./CHANGELOG.md)

Credit: This repo is based on [@manniL](https://github.com/manniL)'s awesome [Nuxt SVG loader](https://www.npmjs.com/package/nuxt-svg-loader).

## Features

- Generate responsive srcsets for your Nuxt.js projects
- Re-size images, compress, convert format, rename and generate placeholders
- Uses [responsive-loader](https://github.com/herrstucki/responsive-loader)
- Uses [Sharp](https://github.com/lovell/sharp/) for fast image processing ⚡
- Fully configurable

## Setup

- Add `nuxt-responsive-loader` as a dependency using yarn or npm
- Add `nuxt-responsive-loader` to the `modules` section of `nuxt.config.js`

```js
{
  modules: ['nuxt-responsive-loader']
}
```

- Add your images to the `assets` directory
- Now you can use the generated srcsets in your templates like so:

```
<template>
  <img :srcset="require('~/assets/nuxt.jpg').srcSet" />
</template>
```

This HTML will be the rendered:

```
<img srcset="/_nuxt/img/2b88a85-640.jpg 640w, /_nuxt/img/1fff45c-750.jpg 750w, /_nuxt/img/6717911-860.jpg 860w, /_nuxt/img/f9f19bf-970.jpg 970w, /_nuxt/img/c0ceb80-1080.jpg 1080w">
```

- The browser will only request the image which matches the current screen size
- Learn more by reading [Responsive Images: If you’re just changing resolutions, use srcset.](https://css-tricks.com/responsive-images-youre-just-changing-resolutions-use-srcset/)

(Tip: if you're using the [nuxt-svg-loader](https://www.npmjs.com/package/nuxt-svg-loader) module, make sure to place it _before_ the nuxt-responsive-loader in the `modules` section of `nuxt.config.js`)

## Configuration

The plugin will work seamlessly out of the box and will use these settings:

```js
{
  name: 'img/[hash:7]-[width].[ext]'
  min: 640 // minimum image width generated
  max: 1080 // maximum image width generated
  steps: 5 // five sizes per image will be generated
  placeholder: false // no placeholder will be generated
  quality: 65 // images are compressed with medium quality
  adapter: require('responsive-loader/sharp')
}
```

If you want to configure the underlying loader, you can do that easily as well. (All options available [here](https://github.com/herrstucki/responsive-loader))

```js
// file: nuxt.config.js

export default {
  // ...
  // Specify your options as a responsiveLoader object
  responsiveLoader: {
    name: 'img/hello-world-[width].[ext]',
    sizes: [200, 500],
    format: 'png',
    adapter: require('responsive-loader/sharp'), // Recommended. Alternatively, leave adapter undefined and add JIMP to your project.
    placeholder: true
  }
}
```

## Examples

- Compressing images to reduce file size:

```js
// file: nuxt.config.js

export default {
  // ...
  // Specify your options as a responsiveLoader object
  responsiveLoader: {
    name: 'img/[hash:7]-[width].[ext]',
    adapter: require('responsive-loader/sharp'),
    quality: 65
  }
}
```

```html
<!-- file: index.vue -->
<template>
  <img :src="require('~/assets/nuxt.jpg').src" />
</template>
```

- Generating placeholders for usage with blur-up technique:

```js
// file: nuxt.config.js

export default {
  // ...
  // Specify your options as a responsiveLoader object
  responsiveLoader: {
    name: 'img/[hash:7]-[width].[ext]',
    adapter: require('responsive-loader/sharp'),
    placeholder: true
  }
}
```

```html
<!-- file: index.vue -->
<template>
  <img :src="require('~/assets/nuxt.jpg').placeholder" data-src="..." />
</template>
```

- Converting image files to `.png`:

```js
// file: nuxt.config.js

export default {
  // ...
  // Specify your options as a responsiveLoader object
  responsiveLoader: {
    name: 'img/[hash:7]-[width].[ext]',
    format: 'png',
    adapter: require('responsive-loader/sharp')
  }
}
```

## Development

- Clone this repository
- Install dependencies using `yarn install` or `npm install`
- Start development server using `npm run dev`

## License

MIT


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## 1.0.2

- Use Sharp loader by default, instead of JIMP


test/module.test.js
const path = require('path')
const klawSync = require('klaw-sync')
const { Nuxt, Builder, Generator } = require('nuxt-edge')

jest.setTimeout(60 * 1000)

const getRelativePath = fileObj => path.relative(__dirname, fileObj.path)

const noJS = item => !/\.js/.test(item)

describe('Default options', () => {
  let nuxt

  beforeAll(async () => {
    nuxt = new Nuxt(require('./fixture/configs/default'))
    await nuxt.ready()
    const builder = new Builder(nuxt)
    await builder.build()
    const generator = new Generator(nuxt)
    await generator.generate({ build: false })
  })

  test('build files contain srcset (.nuxt)', () => {
    const buildFiles = klawSync(nuxt.options.buildDir).map(getRelativePath)
    expect(buildFiles).toContain('../.nuxt/dist/client/img/f04fe61-640.jpg')
    expect(buildFiles).toContain('../.nuxt/dist/client/img/04676df-750.jpg')
    expect(buildFiles).toContain('../.nuxt/dist/client/img/5f9177e-860.jpg')
    expect(buildFiles).toContain('../.nuxt/dist/client/img/b7fb524-970.jpg')
    expect(buildFiles).toContain('../.nuxt/dist/client/img/de5f3f2-1080.jpg')
    expect(buildFiles.filter(noJS)).toMatchSnapshot()
  })

  test('generated files contain srcset (dist)', () => {
    const generateFiles = klawSync(nuxt.options.generate.dir).map(
      getRelativePath
    )
    expect(generateFiles).toContain('../dist/_nuxt/img/f04fe61-640.jpg')
    expect(generateFiles).toContain('../dist/_nuxt/img/04676df-750.jpg')
    expect(generateFiles).toContain('../dist/_nuxt/img/5f9177e-860.jpg')
    expect(generateFiles).toContain('../dist/_nuxt/img/b7fb524-970.jpg')
    expect(generateFiles).toContain('../dist/_nuxt/img/de5f3f2-1080.jpg')
    expect(generateFiles.filter(noJS)).toMatchSnapshot()
  })

  test('HTML uses responsive srcset', async () => {
    const { html } = await nuxt.renderRoute('/')
    expect(html).toContain(
      [
        '<img srcset="',
        '/_nuxt/img/f04fe61-640.jpg 640w,',
        '/_nuxt/img/04676df-750.jpg 750w,',
        '/_nuxt/img/5f9177e-860.jpg 860w,',
        '/_nuxt/img/b7fb524-970.jpg 970w,',
        '/_nuxt/img/de5f3f2-1080.jpg 1080w',
        '">'
      ].join('')
    )
    expect(html).toMatchSnapshot()
  })

  afterAll(async () => {
    await nuxt.close()
  })
})

describe('Custom options, with Jimp', () => {
  let nuxt

  beforeAll(async () => {
    nuxt = new Nuxt(require('./fixture/configs/custom'))
    await nuxt.ready()
    const builder = new Builder(nuxt)
    await builder.build()
    const generator = new Generator(nuxt)
    await generator.generate({ build: false })
  })

  test('responsive srcset reflects custom options', async () => {
    expect(nuxt.renderer).toBeDefined()
    const { html } = await nuxt.renderRoute('/')
    expect(html).toContain(
      [
        '<img srcset="',
        '/_nuxt/img/hello-world-200.png 200w,',
        '/_nuxt/img/hello-world-500.png 500w',
        '">'
      ].join('')
    )
    expect(html).toMatchSnapshot()
    await nuxt.close()
  })

  afterAll(async () => {
    await nuxt.close()
  })
})


test/fixture/configs/custom.js
const { resolve } = require('path')

module.exports = {
  rootDir: resolve(__dirname, '../../../'),
  srcDir: resolve(__dirname, '../'),
  render: {
    resourceHints: false
  },
  modules: ['@@'],
  build: {
    filenames: {
      app: '[name].js',
      chunk: '[name].js'
    }
  },
  responsiveLoader: {
    name: 'img/hello-world-[width].[ext]',
    sizes: [200, 500],
    format: 'png'
  }
}


lib/module.js
const DEFAULT_RESPONSIVE_LOADER_OPTIONS = {
  name: 'img/[hash:7]-[width].[ext]',
  min: 640,
  max: 1080,
  steps: 5,
  placeholder: false,
  quality: 65,
  adapter: require('responsive-loader/sharp')
}

const setupResponsiveLoader = options => config => {
  /* Find the webpack rule which is responsible for images */
  const imageRulePredicate = rule => {
    return (
      rule.test.test('.png') &&
      rule.test.test('.jpg') &&
      rule.test.test('.gif') &&
      rule.test.test('.webp')
    )
  }
  const imageRule = config.module.rules.find(imageRulePredicate)

  if (imageRule) {
    /* Update the rule so it's no longer responsible for png/jpg/gif/webp files */
    if (imageRule.test.test('.svg')) {
      /* If applicable, keep its responsibility for svg files */
      imageRule.test = /\.(svg)$/i
    } else {
      /* Else, remove all its responsibility */
      imageRule.test = /(?!)/
    }
  }

  /* Add a new rule which is responsible for png/jpg/gif/webp files */
  config.module.rules.push({
    test: /\.(png|jpe?g|gif|webp)$/,
    loader: 'responsive-loader',
    options: options || DEFAULT_RESPONSIVE_LOADER_OPTIONS
  })
}

export default function nuxtResponsiveLoader() {
  const { responsiveLoader: options } = this.options
  this.extendBuild(setupResponsiveLoader(options))
}

module.exports.meta = require('../package.json')
